Gstreamer: transition to meson build

zandrey opened the pull request:
OE-Core has transferred gstreamer1.0 plugins to use meson build instead of autotools. This PR addresses the same for i.MX forks maintained in this layer.

@thochstein: There are several patches that should be picked up and merged into respective forks, they are targeting NXP-specific parts of plugin forks in the meson area.

-- andrey

otavio commented:
@thochstein could you review and validate those?

thochstein commented:
I'll try to get this done by the end of the week.

thochstein commented:
Hi guys, I've asked our team to take a look at this and provide feedback. They'll be back from Lunar New Year next week.

zandrey commented:
@thochstein Have you had a chance to get any update from the team regarding this PR?

Just as note: One patch from 53d6ffc0fddf0d4bedb20e194dbdd20f8f3f9ebf is already in the [oe-core] upstream [86511f94c992eca945607b864e935c34fcc146ab](http://git.yoctoproject.org/cgit/cgit.cgi/poky/commit/?id=86511f94c992eca945607b864e935c34fcc146ab), but since we keep the fork here (and all patches as well) - we'd need to have it in the layer anyway.

-- andrey

thochstein commented:
@zandrey Sorry that progress is slow. The team is in China and affected by the coronavirus. Your changes look extensive and the team may need to ramp up on meson to be able to provide a useful review. We'll know more in the next couple of days.

zandrey commented:
@thochstein It's OK, totally understand... If there is some clarification needed from my side - just let me know.

-- andrey

petegriffin commented:
@zandrey does this work ok for you? I tried building this branch and on imx8m-evk board I get SIGSEGV after a few frames on couple of different pipelines I've tried.

e.g. 

root@imx8mqevk:~# gst-launch-1.0 -v souphttpsrc location=https://storage.googleapis.com/wvmedia/clear/h264/tears/tears_uhd.mpd name=source ! application/dash+xml ! dashdemux name=dd0 
dd0.audio_00 ! audio/x-m4a ! fakesink dd0.video_00 ! video/quicktime ! queue ! video/quicktime ! qtdemux name=d1 ! queue ! video/x-h264 ! h264parse ! vpudec  ! waylandsink
Setting pipeline to PAUSED ...
====== VPUDEC: 4.5.1 build on Feb 11 2020 10:10:42. ======
        wrapper: 3.0.0 (VPUWRAPPER_ARM64_LINUX Build on Feb 10 2020 14:14:35)
        vpulib: 1.1.1
        firmware: 1.1.1.0
Pipeline is PREROLLING ...
Got context from element 'source': gst.soup.session=context, session=(SoupSession)NULL, force=(boolean)false;
/GstPipeline:pipeline0/GstCapsFilter:capsfilter0.GstPad:src: caps = application/dash+xml
/GstPipeline:pipeline0/GstDashDemux:dd0.GstPad:sink: caps = application/dash+xml
/GstPipeline:pipeline0/GstCapsFilter:capsfilter5: caps = audio/x-m4a
/GstCapsFilter:capsfilter6: caps = video/quicktime
/GstPipeline:pipeline0/GstCapsFilter:capsfilter7: caps = video/quicktime
/GstPipeline:pipeline0/GstCapsFilter:capsfilter7.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue0.GstPad:sink: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstCapsFilter:capsfilter7.GstPad:sink: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue0.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue0.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue0.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue0.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstCapsFilter:capsfilter3.GstPad:src: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQTDemux:d1.GstPad:sink: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstCapsFilter:capsfilter3.GstPad:sink: caps = video/quicktime, width=(int)3840, height=(int)1714, framerate=(fraction)24/1
/GstPipeline:pipeline0/GstQueue:queue1.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffer)1
/GstPipeline:pipeline0/GstQueue:queue1.GstPad:src: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffer)01
/GstPipeline:pipeline0/GstCapsFilter:capsfilter4.GstPad:src: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data1
/GstPipeline:pipeline0/GstH264Parse:h264parse0.GstPad:src: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(e
/GstPipeline:pipeline0/GstVpuDec:vpudec0.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffee
/GstPipeline:pipeline0/GstH264Parse:h264parse0.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=1
/GstPipeline:pipeline0/GstCapsFilter:capsfilter4.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_dat1
/GstPipeline:pipeline0/GstQTDemux:d1.GstPad:video_0: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffer1
/GstPipeline:pipeline0/GstQueue:queue1.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffer)1
/GstPipeline:pipeline0/GstQueue:queue1.GstPad:src: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=(buffer)01
/GstPipeline:pipeline0/GstCapsFilter:capsfilter4.GstPad:src: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data1
/GstPipeline:pipeline0/GstH264Parse:h264parse0.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_data=1
/GstPipeline:pipeline0/GstCapsFilter:capsfilter4.GstPad:sink: caps = video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)5.1, profile=(string)high, codec_dat1
/GstPipeline:pipeline0/GstVpuDec:vpudec0.GstPad:src: caps = video/x-raw, format=(string)NV12, width=(int)3840, height=(int)1728, interlace-mode=(string)progressive, multiview-mode=(s1
/GstPipeline:pipeline0/GstWaylandSink:waylandsink0.GstPad:sink: caps = video/x-raw, format=(string)NV12, width=(int)3840, height=(int)1728, interlace-mode=(string)progressive, multiv1
Pipeline is PREROLLED ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Caught SIGSEGV
#0  0x0000ffffad6e43dc in poll () from /lib/libc.so.6
#1  0x0000ffffad8057e0 in ?? () from /usr/lib/libglib-2.0.so.0
#2  0x0000ffffad805c38 in g_main_loop_run () from /usr/lib/libglib-2.0.so.0
#3  0x0000ffffad99f614 in gst_bus_poll () from /usr/lib/libgstreamer-1.0.so.0
#4  0x0000000000404858 in ?? ()
#5  0x00000000004036a0 in ?? ()
#6  0x0000ffffad643010 in __libc_start_main () from /lib/libc.so.6
#7  0x0000000000403d38 in ?? ()
Spinning.  Please run 'gdb gst-launch-1.0 3963' to continue debugging, Ctrl-C to quit, or Ctrl-\ to dump core.

and

root@imx8mqevk:~# gst-play-1.0 /FPS_test_1080p25_L4.1.mkv
Press 'k' to see a list of keyboard shortcuts.
Now playing /FPS_test_1080p25_L4.1.mkv
Prerolling...
====== AIUR: 4.5.1 build on Feb 11 2020 10:10:25. ======
	Core: MKVPARSER_01.08.02  build on Feb 15 2019 09:59:27
 file: /usr/lib/imx-mm/parser/lib_mkv_parser_arm_elinux.so.3.1
------------------------
    Track 00 [video_0] Enabled
	Duration: 0:10:05.040000000
	Language: eng
    Mime:
	video/x-h264, parsed=(boolean)true, alignment=(string)au, stream-format=(string)avc, width=(int)1920, height=(int)1080, framerate=(fraction)25/1, codec_data=(buffer)01640029ffe1002a67640029acd900780227e5c05a808080a000000300200000065ccc0002625a0000e4e1d249803c60c65801000668e938233c8f 
------------------------
====== VPUDEC: 4.5.1 build on Feb 11 2020 10:10:42. ======
	wrapper: 3.0.0 (VPUWRAPPER_ARM64_LINUX Build on Feb 10 2020 14:14:35)
	vpulib: 1.1.1
	firmware: 1.1.1.65535
------------------------
    Track 01 [audio_0] Enabled
	Duration: 0:10:05.040000000
	Language: eng
    Mime:
	audio/mpeg, mpegversion=(int)1, channels=(int)2, rate=(int)48000, bitrate=(int)34181 
------------------------

====== BEEP: 4.5.1 build on Feb 11 2020 10:10:31. ======
	Core: MP3 decoder Wrapper  build on Jan 11 2018 10:20:25
 file: /usr/lib/imx-mm/audio-codec/wrap/lib_mp3d_wrap_arm_elinux.so.3
CODEC: BLN_MAD-MMCODECS_MP3D_ARM_02.13.01_ARMV8  build on Jan 11 2018 10:05:45.
Segmentation fault5.0

petegriffin commented:
fyi I just backed out these changes, and reverted meta-openembedded to dcb8c50af47a14dfd0c03097df92b2b8c4782b45 and openembedded-core to b6d4150f9c74f25a4022a3fa0bd489a8e85deb77 (i.e. prior to meson updates). Rebuilt and the pipelines above are working correctly again. So it seems it is definitely a regression with these patches.

zandrey commented:
@petegriffin Thanks for the input! This is actually what've somehow slipped through my verification round. BTW, I'm using i.MX8M Mini EVK for all verifications here.

Initially, I've concentrated on the camera module live view and it is operable with latest GStreamer update. I did the verification now, and following command is working for me:
`gst-launch-1.0 v4l2src device=/dev/video0 ! video/x-raw,width=1920,height=1080 ! glimagesink`

However, I've also figured out that video encoding and decoding are now broken. I do know that at least video encoding was working before, but when I tried it now - it looks like it is not operable anymore. 

Following command fails for me (video encoding):
`gst-launch-1.0 v4l2src device=/dev/video0 ! video/x-raw,width=1920,height=1080 ! vpuenc_h264 ! mpegtsmux ! filesink location=/home/root/test1.ts`

Following is the output from dmesg:
```
[ 3337.045691] alloc_contig_range: [83700, 83af5) PFNs busy
[ 3337.051958] alloc_contig_range: [83800, 83bf5) PFNs busy
[ 3337.057990] alloc_contig_range: [83800, 83cf5) PFNs busy
[ 3337.064061] alloc_contig_range: [83a00, 83df5) PFNs busy
[ 3337.071690] alloc_contig_range: [83700, 83af5) PFNs busy
[ 3337.083091] alloc_contig_range: [83700, 83af5) PFNs busy
[ 3337.091369] alloc_contig_range: [83700, 83af5) PFNs busy
[ 3337.110456] ov5640_mipi 2-003c: s_stream: 1
[ 3337.374764] skip frame 1
[ 3337.412462] alloc_contig_range: [83900, 83afe) PFNs busy
[ 3337.421848] alloc_contig_range: [83a00, 83aff) PFNs busy
[ 3337.436261] ov5640_mipi 2-003c: s_stream: 0
```

As for video decoding case - it is also failing when following command is executed:
`gst-play-1.0 ./ToS-4k-1920.mov`

I also receive a following output from dmesg:
```
[ 2610.737346] alloc_contig_range: [82a00, 83a01) PFNs busy
[ 2610.743801] alloc_contig_range: [82a00, 83b01) PFNs busy
[ 2610.750636] alloc_contig_range: [82c00, 83c01) PFNs busy
[ 2610.756754] alloc_contig_range: [82c00, 83d01) PFNs busy
[ 2610.762839] alloc_contig_range: [82c00, 83e01) PFNs busy
[ 2610.768949] alloc_contig_range: [82c00, 83f01) PFNs busy
[ 2610.775356] alloc_contig_range: [83000, 84001) PFNs busy
[ 2610.782485] alloc_contig_range: [83000, 84101) PFNs busy
[ 2610.789567] alloc_contig_range: [83000, 84201) PFNs busy
[ 2610.796575] alloc_contig_range: [83000, 84301) PFNs busy
[ 2610.870068] imx-sdma 302c0000.dma-controller: sdma firmware not ready!
```

From the dmesg output I can only assume there is an issue with dma buffers that are allocated for Hantro, and are potentially not propagated correct to the GStreamer.

@thochstein Can you please have a look here and maybe advise what could be the issue? Was there something I forgot with this PR, which is imx-specific? Maybe you already have some additional information from your team?

@petegriffin Is there a chance you can verify the Camera live view in your setup? In addition, since there were several other updates - are we absolutely certain that this regression came with this PR?

-- andrey

petegriffin commented:
@zandrey Unfortunately I don't have any camera sensors for imx8m-evk board so I can't run that test (do you have a link to the camera board you're using..maybe I could buy one).

I'm pretty sure the regression came with this pull, as in my now working build I only git reverted these meson changes rather than rolling back whole meta-freescale repo (See my branch here https://github.com/petegriffin/meta-freescale/commits/meson-oemcrypto).

zandrey commented:
@petegriffin The Camera module I used for validation is the official [MIPI-CSI module from NXP with OV5640 sensor](https://www.nxp.com/part/MINISASTOCSI#/)

This module works out of the box with i.MX8M Mini EVK, and should also be supported by i.MX8M EVK.

I would try also to drop all my changes and verify if the video codec is operable again. So far I was not able to identify anything obviously wrong with changes in this PR... I was mainly following NXP modifications they introduced to Makefiles in their fork, and it can be that I missed a part which might be essential.

Anyways, thanks for reporting this and for validation! I'll post an update here once I get more results.

-- andrey

zandrey commented:
I took a bit of digging, and figured out that the issue is actually in `imx-vpuwrap` package, which was not changed in this PR.

Function that SegFaults is:
```
VpuDecRetCode VPU_DecOutFrameDisplayed(VpuDecHandle InHandle, VpuFrameBuffer* pInFrameBuf)
```
and the parameter `pInFrameBuf` is passed as _null_. I tried to trace it down, but the code is a bit convoluted, so I'm progressing here quite slowly.

Neither Decoder nor Encoder are currently operable, only those plugins that are not requiring Hantro are OK (Camera Live View is operable).

I suspect it is maybe some conditional compilation defines that were left out in meson files, which were present in Makefiles before the conversion.

@thochstein Can you please address this further to your team (if they are available)? I don't think I could progress faster here without some guidance from those who knows the component quite well... It might be something trivial that was left out, but for me to find this - might be a huge work. But I would still continue to look into the root cause further.

@petegriffin If it would be a big trouble, can you please get some logs from the build that works for you? I have now some issues to revert the state back, and would appreciate if you can get some additional info posted here. You would need to enable additional logging from VPU Wrapper by creating a file with a content: `echo 1 > /etc/vpu_log_level` and run the `gplay-1.0`. With this, you would have extra info from the Wrapper that I can analyze to compare good and bad cases. 

Thanks a lot!

-- andrey

thochstein commented on recipes-multimedia/gstreamer/gstreamer1.0-plugins-good_1.16.imx.bb:
```diff
@@ -1,81 +1,77 @@
-PACKAGECONFIG[soup]       = "--enable-soup,--disable-soup,libsoup-2.4"
-PACKAGECONFIG[speex]      = "--enable-speex,--disable-speex,speex"
-PACKAGECONFIG[taglib]     = "--enable-taglib,--disable-taglib,taglib"
-PACKAGECONFIG[v4l2]       = "--enable-gst_v4l2 --enable-v4l2-probe,--disable-gst_v4l2,libdrm"
```
Looks like the libdrm dependency was accidentally removed.

zandrey replied:
Nice catch! This was also dropped upstream, therefore I missed this part out. Would ask the author from upstream patch if that was intended.

For this fork - I would bring it back, I've already tested that locally and build is fine.

zandrey replied:
Just realized: this dependency was never in the upstream [commit 41c498bded](http://lists.openembedded.org/pipermail/openembedded-core/2020-January/291294.html)

@thochstein can you please tell if this is really required in this fork?

thochstein commented on recipes-multimedia/gstreamer/gstreamer1.0-plugins-good_1.16.imx.bb:
```diff
@@ -1,81 +1,77 @@
+                    file://gst/replaygain/rganalysis.c;beginline=1;endline=23;md5=b60ebefd5b2f5a8e0cab6bfee391a5fe"
 
-inherit gettext
+DEPENDS += "gstreamer1.0-plugins-base libcap zlib imx-gst1.0-plugin"
```
Just want to make sure that adding a dependency on `imx-gst1.0-plugin` is intentional.

thochstein commented on recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_1.16.imx.bb:
```diff
@@ -53,118 +47,122 @@ PACKAGECONFIG ??= " \
     bz2 curl dash dtls hls rsvg sbc smoothstreaming sndfile ttml uvch264 webp \
```
Should we add `closedcaption` to align with upstream?

zandrey replied:
This would also be added in the next version here, I've left it out since it was not in the old package config.

zandrey commented:
I've finally figured out the issue with playback and decoder wrapper: this was due to the fact that during the upgrade I've forgot to add ION allocator to the meson build (it was only present in autotool part).

All 3 commits are updated, rebased to latest master and ready for further validations.

I've validated the Camera Live View, Video Capture/Encode and Video Decoding functionalities: it is all operable on i.MX8M Mini EVK I have on hands.

@petegriffin can you please try the updated state of this PR and let me know if you still have playback issues?

@thochstein I've updated patches to the fork, you can send those new versions to your team for further validation. I've also resolved all your comments above and addressed them in the updated version.

petegriffin commented:
@zandrey apologies I've been out of office. Let me try a build with your latest patches and get back to you. Also thank you for your efforts in debugging this :) Will be good to have a working ToT build again for meta-freescale

zandrey commented:
Just a small update: I've pushed also the update for `plugins-ugly` and `rtsp-server` components so they are also built with meson. Those packages are not very often in use, but this brings a consistency for the entire _gstreamer_ build now.

@petegriffin Thanks a lot for responding here!
Your validation here would be really appreciated! I currently do not posses the HW you reported the problem on, and if it turns out it works for you then I'd say it should be fine to merge it in.

-- andrey

dkomljenovic2502 commented:
Hi,
just my 2 cents,
I tested this pull req with yocto, and custom imx8mm based board, and after the latest patch update, this seems very nice for me, no problems building the yocto project, and we are able to start the stream without problems. 

BR, DK

zandrey commented:
Just another update here to this PR: I've also ported the `gstreamer1.0` recipe to meson build system. This should conclude the migration of the GStreamer NXP fork and all recipes are now adhere to the `master` branch of oe-core.

I've tested video playback (h.264 and h.265) and live camera view - it is all operable.

@dkomljenovic2502 Thanks a lot for your verification here! The more test coverage we see here for this PR - the better it is! Your report is really appreciated. 

-- andrey
